xdcr.go

```diff
@@ -0,0 +1,245 @@
+// Copyright 2024-Present Couchbase, Inc.
+//
+// Use of this software is governed by the Business Source License included
+// in the file licenses/BSL-Couchbase.txt.  As of the Change Date specified
+// in that file, in accordance with the Business Source License, use of this
+// software will be governed by the Apache License, Version 2.0, included in
+// the file licenses/APL2.txt.
+
+package rosmar
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"sync/atomic"
+
+	sgbucket "github.com/couchbase/sg-bucket"
+)
+
+// XDCR implements a XDCR bucket to bucket setup within rosmar.
+type XDCR struct {
+	terminator              chan bool
+	fromBucketCollectionIDs map[uint32]sgbucket.DataStoreName
+	fromBucket              *Bucket
+	toBucket                *Bucket
+	replicationID           string
+	docsWritten             atomic.Uint64
+	docsFiltered            atomic.Uint64
+	errorCount              atomic.Uint64
+}
+
+// NewXDCR creates an instance of XDCR backed by rosmar. This is not started until Start is called.
+func NewXDCR(_ context.Context, fromBucket, toBucket *Bucket, opts sgbucket.XDCROptions) (*XDCR, error) {
+	if opts.Mobile != sgbucket.XDCRMobileOn {
+		return nil, errors.New("Only sgbucket.XDCRMobileOn is supported in rosmar")
+	}
+	return &XDCR{
+		fromBucket:              fromBucket,
+		toBucket:                toBucket,
+		replicationID:           fmt.Sprintf("%s-%s", fromBucket.GetName(), toBucket.GetName()),
+		fromBucketCollectionIDs: map[uint32]sgbucket.DataStoreName{},
+		terminator:              make(chan bool),
+	}, nil
+
+}
+
+// getFromBucketCollectionName returns the collection name for a given collection ID in the from bucket.
+func (r *XDCR) getFromBucketCollectionName(collectionID uint32) (sgbucket.DataStoreName, error) {
+	dsName, ok := r.fromBucketCollectionIDs[collectionID]
```

Comment:
Since we need to know the set of fromBucket collections when we start the DCP feed anyway, I think populating fromBucketCollectionIDs on start instead of lazily will simplify things and avoid potential races accessing fromBucketCollectionIDs.

Reply:
I made a precondition of this that the DataStores are computed when XDCR.Start is called, and that there have to be matching collection names.

We could relax this in a future PR but I don't think this is necessary for testing at this time.